pkg/gateway/model_build_targets.go

```diff
@@ -120,20 +120,43 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
 			portsAnnotations := strings.Split(serviceExport.ObjectMeta.Annotations[portAnnotationsKey], ",")
 
 			for _, portAnnotation := range portsAnnotations {
-				definedPort, err := strconv.ParseInt(portAnnotation, 10, 64)
+				definedPort, err := strconv.ParseInt(portAnnotation, 10, 32)
 				if err != nil {
 					glog.V(6).Infof("Failed to read Annotations/Port:%s, err:%s", serviceExport.ObjectMeta.Annotations[portAnnotationsKey], err)
 				} else {
-					definedPorts = append(definedPorts, definedPort)
+					definedPorts[int32(definedPort)] = struct{}{}
 				}
 			}
 			glog.V(6).Infof("Build Targets - portAnnotations: %v", definedPorts)
 		}
-	} else if tg.ByBackendRef {
-		definedPorts = []int64{t.port}
+	} else if tg.ByBackendRef && t.port != undefinedPort {
+		definedPorts[t.port] = struct{}{}
 	}
+
+	// A service port MUST have a name if there are multiple ports exposed from a service.
+	// Therefore, if a port is named, endpoint port is only relevant if it has the same name.
+	//
+	// If a service port is unnamed, it MUST be the only port that is exposed from a service.
+	// In this case, as long as the service port is matching with backendRef/annotations,
+	// we can consider all endpoints valid.
+
+	servicePortNames := make(map[string]struct{})
+	skipMatch := false
+
+	for _, port := range svc.Spec.Ports {
+		if _, ok := definedPorts[port.Port]; ok {
+			if port.Name != "" {
+				servicePortNames[port.Name] = struct{}{}
+			} else {
+				// Unnamed, consider all endpoints valid
+				skipMatch = true
+			}
+		}
```

Comment:
For httproute backendref k8sService scenario,
Say, one k8sService have only one unnamed (targetPort 8080, port 80)
And the httproute backendref have a port 81,
In that case we create a lattice target for port 8080, is that expected behavior